AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMCs/StormTracker.dfy

```diff
@@ -177,23 +193,104 @@ module {:options "/functionSyntax:4" }  StormTracker {
         INT64_MAX_LIMIT as Types.PositiveLong
     }
 
+    lemma ExpiryTimeInSecondsCanBeConvertedToMilliseconds()
+      ensures INT64_MAX_LIMIT / 1000 == 0x20C49BA5E353F7
+    {}
+
+    predicate GracePeriod?(result: Types.GetCacheEntryOutput, now : Types.PositiveLong)
+      : (output: bool)
+      reads this`gracePeriod
+      requires now / 1000 <= result.expiryTime
+      ensures
+        output ==>
+          && result.expiryTime < 0x20C49BA5E353F7
+             // The + 1 comes because (now / 1000) * 1000 <= now
+             // Consider (5001 / 1000) <= 5
+             // (5001 / 1000) == 5
+             // but 5 * 1000 < 5001
+          && (result.expiryTime  * 1000) - gracePeriod <= now < (result.expiryTime + 1) * 1000
+    {
+      // This is just showing where this constant comes from
+      // It is so that we know that we can convert
+      // the expiryTime from seconds to milliseconds
+      // This should get fixed sometime before July 29, 31252
+      ExpiryTimeInSecondsCanBeConvertedToMilliseconds();
+      && result.expiryTime < 0x20C49BA5E353F7
+      && (result.expiryTime  * 1000) - gracePeriod <= now
+    }
+
     // If entry is within `grace time` of expiration, then return EmptyFetch once per `grace interval`,
     // and return cached value otherwise
     method CheckInFlight(identifier: seq<uint8>, result: Types.GetCacheEntryOutput, now : Types.PositiveLong)
       returns (output: CacheState)
       modifies this`lastPrune, inFlight
+
+      // Do check an expired result
+      requires now / 1000 <= result.expiryTime
+
+      requires ValidState()
+      ensures ValidState()
+
+      // Only return what was passed in
+      ensures output.Full? ==> output.data == result
+
+      // Never return an expired entry
+      ensures output.Full? ==> now as nat < (result.expiryTime as nat + 1) * 1000
+
+      // EmptyFetch only happens in the grace period
+      ensures output.EmptyFetch? ==> GracePeriod?(result, now)
+
+      // Before the grace period
+      // ==> Full
+      ensures !GracePeriod?(result, now) ==> output.Full?
+
+      // EmptyFetch puts the identifier inFlight
+      ensures output.EmptyFetch? ==>
+                && inFlight.HasKey(identifier)
+                && inFlight.Select(identifier) == now
+
+      // In the grace period,
+      // if fanOut has space and the identifier is not inflight
+      // ==> EmptyFetch
+      ensures
+        && GracePeriod?(result, now)
+        && old(inFlight.Size()) < fanOut as nat
+        && !old(inFlight.HasKey(identifier))
+        ==> output.EmptyFetch?
+
+      // In the grace period,
+      // if fanOut has space and the identifier is inflight but has not exceed the graceInterval
+      // ==> Full
+      ensures
+        && GracePeriod?(result, now)
+        && old(inFlight.Size()) < fanOut as nat
```

Comment:
It seems like this shouldn't be needed. If inFlight were full, we would still return Full.

Reply:
No, this is an edge case :) If inFlight is full at the start, there MAY be a GC and some inflights might go away.